lib/services/connect/middleware/security.js

```diff
@@ -1,156 +1,199 @@
+
+    try {
+
+      if (_this.excluded(req, next)) return;
+
+      if (req.url.substring(0, 1) != '/') req.url = '/' + req.url;
+
+      var parsedUrl = require('url').parse(req.url, true);
+      var query = parsedUrl.query;
+      var path = parsedUrl.pathname;
+      var params = {};
+
+      if (path == '/auth/request-nonce') {
+
+        params.publicKey = _this.happn.services.utils.getFirstMatchingProperty(['publicKey', 'public_key', 'public', 'key', 'public-key'], query);
+
+        return _this.happn.services.security.createAuthenticationNonce(params, function (e, nonce) {
+
+          if (e) return next(e);
+
+          //message, data, error, res, code
+          _this.__respond('nonce generated', nonce, null, res);
+
+        });
+      }
+
+      if (path == '/auth/login') {
+
+        params.username = _this.happn.services.utils.getFirstMatchingProperty(['user', 'username', 'u'], query);
+        params.password = _this.happn.services.utils.getFirstMatchingProperty(['password', 'pwd', 'pass', 'p'], query);
+        params.publicKey = _this.happn.services.utils.getFirstMatchingProperty(['publicKey', 'public_key', 'public', 'key', 'public-key', 'pk'], query);
+        params.digest = _this.happn.services.utils.getFirstMatchingProperty(['digest'], query);
+
+        return _this.happn.services.security.login(params, null, {data:{info:{_local:false}}}, function (e, session) {
+
+          if (e) {
+            if (e.toString() == 'Error: Invalid credentials') return _this.__respond('login failed', null, e, res, 401);
+            return _this.__respond('login failed', null, e, res, 500);
+          }
+
+          _this.__respond('login successful', session.token, null, res);
+        });
+      }
+
+      var session = _this.happn.services.security.sessionFromRequest(req);
+
+      if (!session) return _this.__respondUnauthorized(res, 'invalid token format or null token');
+
+      var url = require('url');
+
+      path = '/@HTTP' + url.parse(req.url).pathname;
+
+      session.type = 0; //stateless session
+
+      _this.happn.services.security.authorize(session, path, req.method.toLowerCase(), function (e, authorized) {
+
+        if (e) {
+
+          if (e.toString().indexOf('AccessDenied') == 0) return _this.__respondForbidden(res, 'unauthorized access to path ' + path);
+          return next(e);
+        }
+
+        if (!authorized) return _this.__respondForbidden(res, 'unauthorized access to path ' + path);
+        req.happn_session = session; //used later if we are rechecking in security
+        next();
+      });
+
+    } catch (e) {
+      next(e);
+    }
+  } else
+    next();
+}
+
+function __respondForbidden(res, message) {
+  var _this = this;
+
+  if (!_this.config.forbiddenResponsePath) {
+    res.writeHead(403, 'unauthorized access', {
+      'content-type': 'text/plain'
+    });
+    return res.end(message);
+  }
+
+  fs.readFile(_this.config.forbiddenResponsePath, function (err, html) {
```

Comment:
I think it would be nicer to have a built-in 403 response if forbiddenResponsePath was not configured, rather than a server error. Same for 401.

Reply:
sorry overlooked that you did just that